src/middleware/aws.rs

```diff
@@ -0,0 +1,98 @@
+use opentelemetry::{
+    global,
+    trace::{SpanKind, Tracer},
+};
+pub use opentelemetry::{trace::Span, KeyValue};
+pub use opentelemetry_semantic_conventions as semcov;
+
+// TODO: Write as macro
+//
+// #[instrument_aws(table_name = "cars", operation = "CreateCar", method = "Post")]
+// fn create_car_in_database() {
+//      info_span_dynamo();
+// }
+
+// Once this scope is closed, all spans inside are closed as well
+#[cfg(any(feature = "aws", feature = "aws_dynamo"))]
+pub fn info_span_dynamo(
+    table_name: &str,
+    operation: &str,
+    method: &str,
+    parent_context: &opentelemetry::Context,
+) -> opentelemetry::global::BoxedSpan {
+    // Spans will be sent to the configured OpenTelemetry exporter
+
+    let tracer = global::tracer("aws_sdk");
+    let mut span = tracer
+        .span_builder("aws_dynamo")
+        .with_attributes(vec![
+            semcov::trace::RPC_METHOD.string(method.to_string()),
+            semcov::trace::RPC_SYSTEM.string("aws-api"),
+            semcov::trace::RPC_SERVICE.string("DynamoDB"),
+        ])
+        .with_kind(SpanKind::Client)
+        .start_with_context(&tracer, parent_context);
```

Comment:
We only really need `parent_context` for sns/firehoose spans because we send them from a different thread in profiles.

Normally we should be able to start this span with `.start` in the current context without any need to do

```rust
let span_context = tracing::Span::current().context();
```

first.

Reply:
I see three ways as to how we could improve this part of the api:

 * we could make `parent_context` optional and then call either `.start` or `.start_with_context` here;
 * or we could skip this call altogether and let the crate user start the span;
 * or we could try to rewrite this method as a macro to avoid passing `None` every time we want to skip this parameter.

Reply:
I just realised that we can't start the span without the tracer which is only available here.

As for the macro, I suggest we leave it for the next step so that we could merge this PR soon.

So, I propose to use the following API:

```rust
let mut span_dynamo_db = telemetry_rust::middleware::aws::info_span_dynamo(
    &table_name,
    "GetProfile",
    "GET",
    None,
);
```

```rust
let mut span_firehose = telemetry_rust::middleware::aws::info_span_firehose(
    &firehose_kinesis_stream_name,
    "SendDataToFirehose",
    "PutRecord",
    Some(&ctx_val),
);
```

Reply:
i.e. accept `Option<&opentelemetry::Context>` to avoid handling context in the service code unless we have to